Newsletters: add 241 (2023-03-08) (#1036)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2023-03-08-newsletter.md

---
title: 'Bitcoin Optech Newsletter #241'
permalink: /en/newsletters/2023/03/08/
name: 2023-03-08-newsletter
slug: 2023-03-08-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter FIXME:harding

## News

- **Alternative design for OP_VAULT:** Greg Sanders [posted][sanders
  vault] to the Bitcoin-Dev mailing list an alternative design for
  providing the features of the `OP_VAULT`/`OP_UNVAULT` proposal (see
  [Newsletter #234][news234 vault]).  His alternative would add three
  opcodes instead of two.  To provide an example:

    - *Alice deposits funds in a vault* by paying a [taproot][topic
      taproot] tree which contains at least two [tapscripts][topic
      tapscript], one with her unvault tapscript and one with her freeze
      tapscript, e.g. `tr(key,{trigger,freeze})`.

>       - The *trigger tapscript* contains her less-trusted authorization

Reviewer: Im still green on this proposal, but above we say "two tapscripts, one with her **unvault** tapscript and one with her **freeze** tapscript" but here we say "trigger" and "freeze" tapscripts

Author: Good catch, thanks!  I rephrased several places to be more consistent, which I think actually simplified things and made them more clear.  Thanks!

        conditions (such as requiring a signature from her hot wallet)
        and an `OP_TRIGGER_FORWARD` opcode.  At the time she creates
        this tapscript, she provides the opcode a *spend delay*
        parameter, e.g. a relative timelock of 1,000 blocks (about 1
        week).

>       - The *freeze tapscript* contains any authorization conditions

Reviewer: Does it make sense to note here that the freeze doesn’t "require" a spend delay? (as I think this is implied when we note later in the writeup that Alice can spend using the freeze tapscript branch if she notices shenanigans)

Author: Add mention about time delays (or lack thereof) to preceding paragraph and also added several other time mentions throughout the text to help clarify this point.  Thanks!

  safe.

  As of this writing, Sanders's proposal had only received limited
  discussion on the mailing list.

- **New Optech Podcast:** the weekly Optech Audio Recap hosted on
  Twitter Spaces is now available as a podcast.  Each episode will be
  available on all popular podcast platforms and on the Optech website
  as a transcript.  For more details, including why we think this is a
  major step forward in Optech's mission to improve Bitcoin technical
  communication, please see our [blog post][podcast post].

## Bitcoin Core PR Review Club

*In this monthly section, we summarize a recent [Bitcoin Core PR Review Club][]
meeting, highlighting some of the important questions and answers.  Click on a
question below to see a summary of the answer from the meeting.*

FIXME:LarryRuane is a PR by FIXME that FIXME.

{% include functions/details-list.md
  q0="FIXME"
  a0="FIXME"
  a0link="https://bitcoincore.reviews/FIXME#l-22"
%}

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

> - [Core Lightning 23.02rc3][] is a release candidate for a new

Reviewer: https://github.com/ElementsProject/lightning/releases/tag/v23.02

      `OP_TAPLEAF_UPDATE_VERIFY` (see [Newsletter #166][news166 tluv]).

    - *Alice completes the unvaulting* by waiting until the relative
      timelock has expired and then spending the unvaulting output,
      choosing the tapleaf with the `OP_FORWARD_OUTPUTS` opcode.  The
      opcode verifies that the spending transaction's output amounts and
      scripts hash to the same commitment Alice made in the previous
      transaction.  In this case, Alice has successfully deposited funds
      to a vault, begun an unvaulting, been forced to wait at least
      1,000 blocks to allow her monitoring programs to verify she really
      did want to spend the funds to the specified outputs, and
      completed the spend.

    - If something goes wrong, *Alice freezes the funds*.  She can do
      this at any time up until an unvaulting is completed by simply
      choosing to spend using the freeze tapscript from the output of
      either the vaulting or initial unvaulting transactions.  Recall
      that Alice explicitly placed the freeze tapscript in the vaulting
      transaction, and note that it was implicitly carried over by the

>       transaction with initiated the unvaulting.

Reviewer 2: ```suggestion
      transaction that initiated the unvaulting.
```

---
title: 'Bitcoin Optech Newsletter #241'
permalink: /en/newsletters/2023/03/08/
name: 2023-03-08-newsletter
slug: 2023-03-08-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter FIXME:harding

## News

- **Alternative design for OP_VAULT:** Greg Sanders [posted][sanders
  vault] to the Bitcoin-Dev mailing list an alternative design for
  providing the features of the `OP_VAULT`/`OP_UNVAULT` proposal (see
  [Newsletter #234][news234 vault]).  His alternative would add three
  opcodes instead of two.  To provide an example:

>     - *Alice deposits funds in a vault* by paying a [taproot][topic
>       taproot] tree which contains at least two [tapscripts][topic
>       tapscript], one with her unvault tapscript and one with her freeze
>       tapscript, e.g. `tr(key,{trigger,freeze})`.

Reviewer 3: Sorry, opinionated on the terminology here. ;)
I would prefer “(taproot) script tree” over “taproot tree”. I would prefer referring to the scripts in the script tree as “leafscripts”, from what I understand “Tapscript” refers to the specific set of Script rules defined by BIP342 to be used in leafscripts with leaf version 0xc0, not a leafscript that uses those rules.

```suggestion
    - *Alice deposits funds in a vault* by paying a P2TR output with a [script tree][topic
      taproot] which contains at least two [leafscripts][topic
      tapscript], one with her unvault leafscript and one with her freeze
      leafscript, e.g. `tr(key,{trigger,freeze})`.
```

Author: > I would prefer “(taproot) script tree” over “taproot tree”.

I'm fine with this.

> I would prefer referring to the scripts in the script tree as “leafscripts”, from what I understand “Tapscript” refers to the specific set of Script rules defined by BIP342 to be used in leafscripts with leaf version 0xc0, not a leafscript that uses those rules.

I can't think of a good argument against this (and, yeah, I tried :-), so I'll make these changes, but since you have a strong opinion about this, would you consider opening a PR to edit https://bitcoinops.org/en/topics/tapscript/ to include a description of the differences between tapscript and leafscript (and adding an alias in the YAML so that we get leafscript in our topics index)?  If you're willing, you can do this at your leisure; it's not blocking on merging this newsletter.  Please CC me if you open a PR as I don't subscribe to all PRs on the repo.

>       - The *trigger tapscript* contains her less-trusted authorization
>         conditions (such as requiring a signature from her hot wallet)
>         and an `OP_TRIGGER_FORWARD` opcode.  At the time she creates
>         this tapscript, she provides the opcode a *spend delay*
>         parameter, e.g. a relative timelock of 1,000 blocks (about 1
>         week).

Reviewer 3: ```suggestion
      - The *trigger leafscript* contains her less-trusted authorization
        conditions (such as requiring a signature from her hot wallet)
        and an `OP_TRIGGER_FORWARD` opcode.  At the time she creates
        this leafscript, she provides the opcode a *spend delay*
        parameter, e.g. a relative timelock of 1,000 blocks (about 1
        week).
```

>       - The *freeze tapscript* contains any authorization conditions
>         Alice wants to specify (including none at all) and
>         an `OP_FORWARD_DESTINATION` opcode.  At the time she creates
>         this tapscript, she also chooses her more-trusted authorization
>         conditions (such as requiring multiple signatures from multiple
>         cold wallets and hardware signing devices).  She provides the
>         opcode a commitment to those conditions in the form of a hash
>         digest.

Reviewer 3: ```suggestion
      - The *freeze leafscript* contains any authorization conditions
        Alice wants to specify (including none at all) and
        an `OP_FORWARD_DESTINATION` opcode.  At the time she creates
        this leafscript, she also chooses her more-trusted authorization
        conditions (such as requiring multiple signatures from multiple
        cold wallets and hardware signing devices).  She provides the
        opcode a commitment to those conditions in the form of a hash
        digest.
```

    - *Alice initiates an unvaulting* by spending the output received to

>       the above taproot tree (using it as an input) and choosing the

Reviewer 3: Since the script tree is only a subset of the spending conditions in a P2TR output, the “received to the taproot tree” feels a bit odd to me. 

```suggestion
      the output with the above script tree (using it as an input) and choosing the
```

>       trigger tapscript.  At this time, she provides two additional

Reviewer 3: ```suggestion
      trigger leafscript.  At this time, she provides two additional
```

      parameters to the `OP_TRIGGER_FORWARD` opcode, the index of the
      output which will receive this input's funds and a hash-based
      commitment to how she wants to be able to spend the funds later.
      The opcode verifies that the indicated output of this transaction

>       pays a taproot tree similar to the one being spent except that the
>       trigger tapscript is replaced with a script using an

Reviewer 3: ```suggestion
      pays a P2TR output with a script tree similar to the one being spent except that the
      trigger leafscript is replaced with a script using an
```

      `OP_CHECKSEQUENCEVERIFY` (CSV) relative delay equal to the delay
      specified previously (e.g., 1000 blocks) and an
      `OP_FORWARD_OUTPUTS` opcode which includes Alice's commitment

>       hash.  The method of reconstructing the taproot tree is similar to
>       an earlier [covenant][topic covenants] proposal,

Reviewer 3: ```suggestion
      hash.  The method of reconstructing the script tree is similar to
      an earlier [covenant][topic covenants] proposal,
```

      `OP_TAPLEAF_UPDATE_VERIFY` (see [Newsletter #166][news166 tluv]).

    - *Alice completes the unvaulting* by waiting until the relative
      timelock has expired and then spending the unvaulting output,
      choosing the tapleaf with the `OP_FORWARD_OUTPUTS` opcode.  The
      opcode verifies that the spending transaction's output amounts and

>       scripts hash to the same commitment Alice made in the previous

Reviewer 3: ```suggestion
      script’s hash to the same commitment Alice made in the previous
```

      transaction.  In this case, Alice has successfully deposited funds
      to a vault, begun an unvaulting, been forced to wait at least
      1,000 blocks to allow her monitoring programs to verify she really
      did want to spend the funds to the specified outputs, and
      completed the spend.

>     - If something goes wrong, *Alice freezes the funds*.  She can do
>       this at any time up until an unvaulting is completed by simply
>       choosing to spend using the freeze tapscript from the output of
>       either the vaulting or initial unvaulting transactions.  Recall
>       that Alice explicitly placed the freeze tapscript in the vaulting
>       transaction, and note that it was implicitly carried over by the
>       transaction with initiated the unvaulting.

Reviewer 3: ```suggestion
    - If something goes wrong, *Alice freezes the funds*.  She can do
      this at any time up until an unvaulting is completed by simply
      choosing to spend using the freeze leafscript from the output of
      either the vaulting or initial unvaulting transactions.  Recall
      that Alice explicitly placed the freeze leafscript in the vaulting
      transaction, and note that it was implicitly carried over by the
      transaction with initiated the unvaulting.
```

>   One of the advantages to users of this approach over the original
>   `OP_VAULT` design is that the freeze tapscript can contain any
>   authorization conditions Alice wants to specify.  In the `OP_VAULT`
>   proposal, anyone knowing the parameters chosen by Alice could spend
>   her funds to the freeze script.  That wasn't a security problem but it
>   could be annoying.  In Sanders's design, Alice could (for example)
>   require a signature from a very lightly protected wallet in order to
>   initiate a freeze---this would perhaps be enough of a burden to
>   prevent most griefing attacks but not enough of a barrier to prevent
>   Alice from quickly freezing her funds in an emergency.

Reviewer 3: ```suggestion
  One of the advantages to users of this approach over the original
  `OP_VAULT` design is that the freeze leafscript can contain any
  authorization conditions Alice wants to specify.  In the `OP_VAULT`
  proposal, anyone knowing the parameters chosen by Alice could spend
  her funds to the freeze script.  That wasn't a security problem but it
  could be annoying.  In Sanders's design, Alice could (for example)
  require a signature from a very lightly protected wallet in order to
  initiate a freeze---this would perhaps be enough of a burden to
  prevent most griefing attacks but not enough of a barrier to prevent
  Alice from quickly freezing her funds in an emergency.
```

  prevent most griefing attacks but not enough of a barrier to prevent
  Alice from quickly freezing her funds in an emergency.

  Several other advantages are aimed at making the consensus-enforced
  [vaulting protocol][topic vaults] easier to understand and verify as
  safe.

  As of this writing, Sanders's proposal had only received limited
  discussion on the mailing list.

- **New Optech Podcast:** the weekly Optech Audio Recap hosted on
  Twitter Spaces is now available as a podcast.  Each episode will be
  available on all popular podcast platforms and on the Optech website
  as a transcript.  For more details, including why we think this is a
  major step forward in Optech's mission to improve Bitcoin technical
  communication, please see our [blog post][podcast post].

## Bitcoin Core PR Review Club

*In this monthly section, we summarize a recent [Bitcoin Core PR Review Club][]
meeting, highlighting some of the important questions and answers.  Click on a
question below to see a summary of the answer from the meeting.*

[bitcoin-inquisition: Activation logic for testing consensus changes][review club bi-16]

> is a PR by Anthony Towns that improves activation and deactivation
> logic for the [Bitcoin Inquisition][] project, which runs on [signet][topic signet].

Reviewer 2: I quite strongly think "improves" is not the right word here. It swaps it out for a different mechanism; they are used in different contexts with very different goals. I don't think there is any notion of one being better than the other.
```suggestion
is a PR by Anthony Towns that adds a new method for activating and deactivating
soft forks in the [Bitcoin Inquisition][] project, designed to be run on [signet][topic signet]
and used for testing.
```

This project was covered in [newsletter #219 bi][].

> While most of the [Bitcoin Inquisition PRs][bi prs] implement specific consensus or
> relay policy changes, this PR modifies the Bitcoin Inquisition framework itself.

Reviewer 2: ```suggestion
```
It's unclear what "the Bitcoin Inquisition framework" would mean.

In contrast to consensus and relay changes on mainnet -- which are difficult
and time-consuming to activate, requiring the careful building of (human)
consensus and an elaborate [soft fork activation][topic soft-fork-activation]
mechanism -- on a test network activating these changes can be streamlined.
The PR also implements a way to deactivate changes that turn out to be buggy
or undesired, which is a major departure from mainnet.
Specifically, this PR replaces [BIP9][] block version bit semantics with what

> are called [Heretical Deployments][].

Reviewer 2: I think the paragraph would flow better if this was the first sentence, rather than the last sentence.

{% include functions/details-list.md
  q0="Why do we want to deploy consensus changes that aren’t merged
      into Bitcoin Core? What problems (if any) are there with merging the
      code into Bitcoin Core, and then testing it on signet afterward?"
  a0="Several reasons were discussed. We can't require mainnet users to upgrade
      the version of Core they're
      running, so even after a bug has been fixed, some users may continue
      running the buggy version. Depending only on regtest makes
      integration testing third-party software more difficult.

>       The threshold of merging consensus changes to a signet-only repository
>       is much lower than merging to Core.
>       Merging changes to Core that aren't yet activated introduces the
>       risk that bugs can leak in a way that affects existing behavior."

Reviewer 2: ```suggestion
      Merging consensus changes to a separate repository is much less risky than merging to Core;
      adding soft fork logic, even if not activated, may introduce bugs that affect existing behavior."
```

      (`DEFINED`, `STARTED`, `LOCKED_IN`, `ACTIVE`, and `FAILED`),
      but with one additional state after `ACTIVE` called `DEACTIVATING`
      (following which is the final state, `ABANDONED`).
      What is the purpose of the `DEACTIVATING` state?"
  a1="It gives users a chance to withdraw funds they might have locked
      into the soft fork. Once the fork is deactivated or replaced, they
      might not be able to spend the funds at all -- even if they're
      anyone-can-spend; that doesn't work if your tx is rejected for
      being non-standard.
      The concern isn't so much the permanent
      loss of the limited signet funds, but rather that the UTXO set
      may become bloated."
  a1link="https://bitcoincore.reviews/bitcoin-inquisition-16#l-92"

  q2="Why does the PR remove `min_activation_height`?"
  a2="We don't need a configurable interval between lock-in and activation
      in the new state model -- with Heretical Deployments, it activates
      automatically at the start of the next 432-block (3 days) state
      machine period (this period is fixed for Heretical Deployment)."
  a2link="https://bitcoincore.reviews/bitcoin-inquisition-16#l-126"

>   q3="Why is Taproot buried in signet?"

Reviewer 2: ```suggestion
  q3="Why is Taproot buried in this PR?"
```

      choosing the tapleaf with the `OP_FORWARD_OUTPUTS` opcode.  The
      opcode verifies that the spending transaction's output amounts and
      script’s hash to the same commitment Alice made in the previous
      transaction.  In this case, Alice has successfully deposited funds
      to a vault, begun an unvaulting, been forced to wait at least
      1,000 blocks to allow her monitoring programs to verify she really
      did want to spend the funds to the specified outputs, and
      completed the spend.

    - If something goes wrong, *Alice freezes the funds*.  She can do
      this at any time from the moment she deposits funds in the vault
      up until an unvaulting is completed.  To freeze funds, she simply
      chooses to spend the freeze leafscript from the output of
      either the vaulting or trigger transactions.  Recall
      that Alice explicitly placed the freeze leafscript in the vaulting
      transaction, and note that it was implicitly carried over by the
      trigger transaction which initiated the unvaulting.

  One of the advantages to users of this approach over the original
  `OP_VAULT` design is that the freeze leafscript can contain any

>   authorization conditions Alice wants to specify.  In the `OP_VAULT`

Reviewer 4: For what it's worth, this was the case with BIP 345 before Sanders made these suggestions. Search for [`optional auth. scriptPubKey](https://github.com/jamesob/bips/blob/jamesob-23-02-opvault/bip-vaults.mediawiki#user-content-Parameters) in the current text of the BIP. Greg's rework doesn't add this in.

Author: Was that the result of some previous change in the design?  In our original coverage of the idea, we linked to a post by AJ where he mentioned that as a concern, see https://bitcoinops.org/en/newsletters/2023/01/18/#proposal-for-new-vault-specific-opcodes where we wrote,

> Towns [notes](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2023-January/021328.html) that [...] any third party who learns the address can also freeze the user’s funds (although they’ll have to pay a transaction fee to do so), creating an inconvenience for the user.

If it was a previous change to the proposal that added the ability to allow requiring arbitrary authorization for freezing, then I think we're technically correct to say the an advantage of Greg's approach over the original approach (as posted to the ML on Jan 9th) is this ability.

However, I also get that it's misleading in the sense that it's not an advantage over what you wanted to ship at the time Greg suggested it.  Would you like us to print a correction in next week's newsletter?
